cmd/admin-trace.go

```diff
@@ -0,0 +1,295 @@
+	NodeName     string       `json:"nodename"`
+	FuncName     string       `json:"funcname"`
+	RequestInfo  requestInfo  `json:"request"`
+	ResponseInfo responseInfo `json:"response"`
+}
+
+// parse Operation name from function name
+func getOpName(fname string) (op string) {
+	res := funcNameRegex.FindStringSubmatch(fname)
+	op = fname
+	if res != nil && len(res) == 2 {
+		op = res[1]
+	}
+	return
+}
+
+// return a struct with minimal trace info.
+func shortTrace(ti madmin.TraceInfo) shortTraceMsg {
+	s := shortTraceMsg{}
+	t := ti.Trace
+	s.NodeName = t.NodeName
+	s.Time = t.ReqInfo.Time
+	if host, ok := t.ReqInfo.Headers["Host"]; ok {
+		s.Host = strings.Join(host, "")
+	}
+	s.Path = t.ReqInfo.Path
+	s.Query = t.ReqInfo.RawQuery
+	s.FuncName = getOpName(t.FuncName)
+	s.StatusCode = t.RespInfo.StatusCode
+	s.StatusMsg = http.StatusText(t.RespInfo.StatusCode)
+
+	return s
+}
+func (s shortTraceMsg) JSON() string {
+	traceJSONBytes, e := json.MarshalIndent(s, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+	return string(traceJSONBytes)
+}
+func (s shortTraceMsg) String() string {
+	var hostStr string
+	if s.Host != "" {
+		hostStr = colorizedNodeName(s.Host)
+	}
+	msg := fmt.Sprintf("%s %s ", s.Time.Format(timeFormat), console.Colorize("FuncName", s.FuncName))
+	msg += fmt.Sprintf("%s%s", hostStr, s.Path)
+	if s.Query != "" {
+		msg += fmt.Sprintf("?%s", s.Query)
+	}
+	msg += console.Colorize("ResponseStatus", fmt.Sprintf("\t%s %s", strconv.Itoa(s.StatusCode), s.StatusMsg))
+	return msg
+}
+
+// colorize node name
+func colorizedNodeName(nodeName string) string {
+	nodeHash := fnv.New32a()
+	nodeHash.Write([]byte(nodeName))
+	nHashSum := nodeHash.Sum32()
+	idx := int(nHashSum) % len(colors)
+	return console.Colorize(fmt.Sprintf("Node%d", colors[idx]), fmt.Sprintf("%s", nodeName))
+}
+
+func (t traceMessage) JSON() string {
+	rqHdrs := make(map[string]string)
+	rspHdrs := make(map[string]string)
+	rq := t.Trace.ReqInfo
+	rs := t.Trace.RespInfo
+	for k, v := range rq.Headers {
+		rqHdrs[k] = strings.Join(v, " ")
+	}
+	for k, v := range rs.Headers {
+		rspHdrs[k] = strings.Join(v, " ")
+	}
+	trc := trace{
+		NodeName: t.Trace.NodeName,
+		FuncName: t.Trace.FuncName,
+		RequestInfo: requestInfo{
+			Time:     rq.Time,
+			Method:   rq.Method,
+			Path:     rq.Path,
+			RawQuery: rq.RawQuery,
+			Body:     string(rq.Body),
+			Headers:  rqHdrs,
+		},
+		ResponseInfo: responseInfo{
+			Time:       rs.Time,
+			Body:       string(rs.Body),
+			Headers:    rspHdrs,
+			StatusCode: rs.StatusCode,
+		},
+	}
+	buf := &bytes.Buffer{}
+	enc := json.NewEncoder(buf)
+	enc.SetIndent("", " ")
+	enc.SetEscapeHTML(false)
+	enc.Encode(trc)
+	// strip off extra newline added by json encoder
+	bufStr := string(buf.Bytes())
+	return strings.TrimSuffix(bufStr, "\n")
+}
+
+func (t traceMessage) String() string {
+	var nodeNameStr string
+	trc := t.Trace
+	if trc.NodeName != "" {
+		nodeNameStr = fmt.Sprintf("%s ", colorizedNodeName(trc.NodeName))
```

Comment:
In the interest of efficient use of memory, we could use [strings.Builder](https://golang.org/pkg/strings/#Builder) type to construct the `String` representation of `traceMessage` value.

Reply:
done.